### PR TITLE
[SPARK-27815][SQL] Predicate pushdown in one pass for cascading joins

### PR DESCRIPTION
**Pull request by yeshengm**
## What changes were proposed in this pull request?

This PR makes the predicate pushdown logic in catalyst optimizer more efficient by unifying two existing rules `PushdownPredicates` and `PushPredicateThroughJoin`. Previously pushing down a predicate for queries such as `Filter(Join(Join(Join)))` requires n steps. This patch essentially reduces this to a single pass.

To make this actually work, we need to unify a few rules such as `CombineFilters`, `PushDownPredicate` and `PushDownPrdicateThroughJoin`. Otherwise cases such as `Filter(Join(Filter(Join)))` still requires several passes to fully push down predicates. This unification is done by composing several partial functions, which makes a minimal code change and can reuse existing UTs.

Results show that this optimization can improve the catalyst optimization time by 16.5%. For queries with more joins, the performance is even better. E.g., for TPC-DS q64, the performance boost is 49.2%.

## How was this patch tested?
Existing UTs + new a UT for the new rule.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -74,7 +73,6 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
         CollapseRepartition,
         CollapseProject,
         CollapseWindow,
-        CombineFilters,
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Let us keep this.

**Comment by gatorsmile** [comment]
LGTM except one comment

**Review by maropu — APPROVED** [review]

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by yeshengm** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by dongjoon-hyun** [comment]
What about updating the JIRA instead since the commit is permanant? @gatorsmile and @cloud-fan .
We can switch both JIRAs.

**Comment by HyukjinKwon** [comment]
K. let me do that. I did this once before as well.

**Comment by HyukjinKwon** [comment]
K. Let me do that.

**Comment by dongjoon-hyun** [comment]
Thanks!

**Pull request closed without merging**